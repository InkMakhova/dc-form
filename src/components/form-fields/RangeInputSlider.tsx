import { JSX, useEffect, useState } from 'react'
import styles from './RangeInputSlider.module.css'
import { RangeInputSliderTypesProps } from './RangeInputSlider.types.ts'
import FormNumberInput from '../form-number-input/FormNumberInput.tsx'
import FormSlider from '../form-slider/FormSlider.tsx'

function RangeInputSlider(props: RangeInputSliderTypesProps): JSX.Element {
  const {
    inputName,
    inputLabel,
    sliderAriaLabel,
    min,
    max,
    step,
    defaultValue
  } = props;

  // State
  const [state, setState] = useState(defaultValue);

  // Handlers
  const increment = () => { setState(prevState => prevState + step) }
  const decrement = () => { setState(prevState => prevState - step) }
  const handleChange = (value: number) => { setState(value) }

  // Update state
  useEffect(() => { setState(defaultValue) }, [defaultValue])

  // Render
  return (
    <div className={styles['range-input-slider']}>
      {/* Input control */}
      <div className={styles['range-input-slider__input']}>
        <FormNumberInput
          name={inputName}
          label={inputLabel}
          min={min}
          max={max}
          step={step}
          value={state}
          onIncrement={increment}
          onDecrement={decrement}
        />
      </div>

      {/* Slider */}
      <div className={styles['range-input-slider__slider']}>
        <FormSlider
          ariaLabel={sliderAriaLabel}
          min={min}
          max={max}
          step={step}
          value={state}
          onChange={handleChange}
        />
      </div>
    </div>
  )
}

export default RangeInputSlider;
