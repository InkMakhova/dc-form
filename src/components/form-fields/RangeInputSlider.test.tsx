import { fireEvent, render, screen } from '@testing-library/react'
import RangeInputSlider from './RangeInputSlider.tsx'
import { expect } from 'vitest';

test('input has default value', () => {
  const { container } = render(
    <RangeInputSlider
      inputName='size'
      inputLabel='Size'
      sliderAriaLabel='Size Slider'
      min={0}
      max={100}
      step={5}
      defaultValue={30}
    />
  );

  const input = container.querySelector('input');
  expect(input).toHaveValue('30');

  const slider = screen.getByLabelText('Size Slider')
  expect(slider).toBeInTheDocument()
});

test('updates value when increment button is clicked', () => {
  const { container } = render(
    <RangeInputSlider
      inputName='size'
      inputLabel='Size'
      sliderAriaLabel='Size Slider'
      min={0}
      max={100}
      step={10}
      defaultValue={30}
    />
  );

  const input = container.querySelector('input');
  const incrementButton = screen.getByLabelText('Increase size')

  fireEvent.click(incrementButton)
  expect(input).toHaveValue('40');
})

test('updates value when decrement button is clicked', () => {
  const { container } = render(
    <RangeInputSlider
      inputName='size'
      inputLabel='Size'
      sliderAriaLabel='Size Slider'
      min={0}
      max={100}
      step={10}
      defaultValue={30}
    />
  );

  const input = container.querySelector('input');
  const decrementButton = screen.getByLabelText('Decrease size')

  fireEvent.click(decrementButton)
  expect(input).toHaveValue('20');
})

test('matches snapshot', () => {
  const { container } = render(
    <RangeInputSlider
      inputName='size'
      inputLabel='Size'
      sliderAriaLabel='Size Slider'
      min={0}
      max={100}
      step={10}
      defaultValue={30}
    />
  )
  expect(container).toMatchSnapshot();
});

