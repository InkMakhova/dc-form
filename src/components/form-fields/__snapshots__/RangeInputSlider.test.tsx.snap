// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot 1`] = `
<div>
  <div
    class="_range-input-slider_963c2d"
  >
    <div
      class="_range-input-slider__input_963c2d"
    >
      <div
        class="_input__container_7e3432"
        data-rac=""
      >
        <label
          class="react-aria-Label"
          for="react-aria-«r1m»"
          id="react-aria-«r1o»"
        >
          Size
        </label>
        <div
          class="_input-number__group-container_9885f5"
          data-rac=""
          role="group"
        >
          <button
            aria-controls="react-aria-«r1m»"
            aria-label="Decrease size"
            aria-labelledby="react-aria-«r1t» react-aria-«r1o»"
            class="
        _button_4e4ded
        _button--small_4e4ded
        _button--secondary-contained_4e4ded
        _button--icon-only_4e4ded
        _button--decrement_4e4ded
      "
            data-rac=""
            data-testid="form-button"
            id="react-aria-«r1t»"
            slot="decrement"
            tabindex="-1"
            type="button"
          >
            <span
              class="
            _button__icon_4e4ded 
            
          "
            >
              <svg
                fill="none"
                height="2"
                viewBox="0 0 14 2"
                width="14"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 1.99921V-0.000671387H13.9992V1.99921H0Z"
                  fill="#1A1A1A"
                />
              </svg>
            </span>
          </button>
          <input
            aria-labelledby="react-aria-«r1o»"
            aria-readonly="true"
            aria-roledescription="Number field"
            autocomplete="off"
            autocorrect="off"
            class="
            _input__field_7e3432 
            _input__field--small_7e3432
            _input-number_9885f5
          "
            data-rac=""
            id="react-aria-«r1m»"
            inputmode="numeric"
            name="size"
            readonly=""
            spellcheck="false"
            tabindex="0"
            title=""
            type="text"
            value="30"
          />
          <button
            aria-controls="react-aria-«r1m»"
            aria-label="Increase size"
            aria-labelledby="react-aria-«r1s» react-aria-«r1o»"
            class="
        _button_4e4ded
        _button--small_4e4ded
        _button--secondary-contained_4e4ded
        _button--icon-only_4e4ded
        _button--increment_4e4ded
      "
            data-rac=""
            data-testid="form-button"
            id="react-aria-«r1s»"
            slot="increment"
            tabindex="-1"
            type="button"
          >
            <span
              class="
            _button__icon_4e4ded 
            
          "
            >
              <svg
                fill="none"
                height="14"
                viewBox="0 0 14 14"
                width="14"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M7.49988 0.499695V6.49969H13.4989V7.49969H7.49988V13.4987H6.49988V7.49969H0.499878V6.49969H6.49988V0.499695H7.49988Z"
                  fill="#1A1A1A"
                  stroke="black"
                />
              </svg>
            </span>
          </button>
        </div>
      </div>
    </div>
    <div
      class="_range-input-slider__slider_963c2d"
    >
      <div
        aria-label="Size Slider"
        class="_slider_f7dc6d"
        data-orientation="horizontal"
        data-rac=""
        id="react-aria-«r22»"
        role="group"
      >
        <div
          class="_slider__track_f7dc6d"
          data-orientation="horizontal"
          data-rac=""
          style="position: relative;"
        >
          <div
            class="_slider__progress_f7dc6d"
            style="width: 30%;"
          />
          <div
            class="_slider__thumb_f7dc6d"
            data-rac=""
            style="position: absolute; left: 30%; transform: translate(-50%, -50%);"
          >
            <div
              style="border: 0px; clip-path: inset(50%); height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
            >
              <input
                aria-describedby=""
                aria-details=""
                aria-labelledby="react-aria-«r22»"
                aria-orientation="horizontal"
                aria-valuetext="30"
                id="react-aria-«r22»-0"
                max="100"
                min="0"
                step="10"
                tabindex="0"
                type="range"
                value="30"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
